pkg/reconciler/workload/resource/resource_controller.go

```diff
@@ -0,0 +1,355 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resource
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/kcp-dev/logicalcluster"
+
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/dynamic"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	"k8s.io/client-go/kubernetes"
+	corelisters "k8s.io/client-go/listers/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/clusters"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	tenancyinformers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/tenancy/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/informer"
+	"github.com/kcp-dev/kcp/pkg/reconciler/workload/namespace"
+)
+
+const controllerName = "kcp-workload-resource-scheduler"
+
+type clusterDiscovery interface {
+	WithCluster(name logicalcluster.Name) discovery.DiscoveryInterface
+}
+
+// NewController returns a new Controller which schedules resources in scheduled namespaces.
+func NewController(
+	dynamicClusterClient dynamic.ClusterInterface,
+	dynamicMetadataClusterClient dynamic.ClusterInterface,
+	clusterDiscoveryClient clusterDiscovery,
+	kubeClusterClient kubernetes.ClusterInterface,
+	workspaceInformer tenancyinformers.ClusterWorkspaceInformer,
+	namespaceInformer coreinformers.NamespaceInformer,
+	pollInterval time.Duration,
+) *Controller {
+	resourceQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-namespace-resource")
+	gvrQueue := workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kcp-namespace-gvr")
+
+	c := &Controller{
+		resourceQueue: resourceQueue,
+		gvrQueue:      gvrQueue,
+
+		dynClient: dynamicClusterClient,
+
+		namespaceLister: namespaceInformer.Lister(),
+
+		kubeClient: kubeClusterClient,
+	}
+
+	namespaceInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: filterNamespace,
+		Handler: cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) { c.enqueueNamespace(obj) },
+			UpdateFunc: func(old, obj interface{}) {
+				oldNS := old.(*corev1.Namespace)
+				newNS := obj.(*corev1.Namespace)
+				if !reflect.DeepEqual(scheduleStateLabels(oldNS.Labels), scheduleStateLabels(newNS.Labels)) {
```

Comment:
ns.Annotations should be used instead of labels as the placement_controller use annotations and not label.
The use of ns,label is accross the board and so needs to be changed in several places.

Reply:
if I am not mistaken, these are different keys: `schedulingv1alpha1.PlacementAnnotationKey` is an annotation, `workloadv1alpha1.InternalClusterResourceStateLabelPrefix` and `namespace.DeprecatedScheduledClusterNamespaceLabel` are about labels.

Reply:
fixed